units/host_allocator.go

```diff
@@ -199,26 +199,35 @@ func (j *hostAllocatorJob) Run(ctx context.Context) {
 	})
 
 	var makespan time.Duration
-	if len(hostsSpawned) != 0 {
-		makespan = distroQueueInfo.ExpectedDuration / time.Duration(len(hostsSpawned))
-	} else if distroQueueInfo.Length > 0 {
-		makespan = distroQueueInfo.ExpectedDuration
-	}
+
+	// ignoring all the tasks that will take longer than the threshold to run,
+	// how long will it take the current fleet of hosts, plus the ones we spawned, to chew through
+	// the scheduled tasks in the queue?
+	scheduledDuration := distroQueueInfo.ExpectedDuration - distroQueueInfo.DurationOverThreshold
+
+	existingCap := time.Duration(nHostsFree) * distroQueueInfo.MaxDurationThreshold
+	spawnCap := time.Duration(len(hostsSpawned)) * distroQueueInfo.MaxDurationThreshold
+
+	makespan = scheduledDuration / (existingCap + spawnCap)
```

Comment:
Could you explain what you're trying to measure here? The makespan variable has units of nothing, since you're dividing a duration by another duration. Should the denominator be some number of hosts?

Reply:
As written, I think you can interpret this as the number of time chunks of length MaxDurationThreshold it would take for (the hosts currently free + the newly-spawned hosts) to work through the backlog of tasks in the queue. If this is 1, everything is working as expected. if this is .5, that means that the queue will be empty in distroQueueInfo.MaxDurationThreshold * .5 minutes, and so on. 

The immediate motivations for me are:
1) figuring out when this number is significantly below one so we can scale back hosts before they become idle
2) Having a better metric of queue health, as alerting on task wait from the task-end-stats report relies on these tasks actually completing.

I'm glad you pointed this out because I had included seconds conversions below, which don't make any sense with the new number. As is, I like the unitless number because it should make it actionable on its own, without having to know MaxDurationThreshold. But I should also have a metric with duration. I'll work one up and re-request when it's ready

Reply:
As written, I think you can interpret this as the number of time chunks of length MaxDurationThreshold it would take for (the hosts currently free + the newly-spawned hosts) to work through the backlog of tasks in the queue. If this is 1, everything is working as expected. if this is .5, that means that the queue will be empty in distroQueueInfo.MaxDurationThreshold * .5 minutes, and so on. 

The immediate motivations for me are:
1) figuring out when this number is significantly below one so we can scale back hosts before they become idle
2) Having a better metric of queue health, as alerting on task wait from the task-end-stats report relies on these tasks actually completing.

I'm glad you pointed this out because I had included seconds conversions below, which don't make any sense with the new number. As is, I like the unitless number because it should make it actionable on its own, without having to know MaxDurationThreshold. But I should also have a metric with duration. I'll work one up and re-request when it's ready